### ISSUE TITLE
Password protected sessions for pre-planned events

### ISSUE THREAD

**Issue by sirkitree** [Frooxius/NeosPublic#351]
Currently we have the ability to have public and private sessions, or only friends. One feature of other major VR social platforms is the ability to have password protected sessions. This allow someone to not have to be friends ahead of time, and not need to use individual invites to a session, giving them just a password instead. Then the session can be listed publicly, and when someone tries to join, must provide the password.

Here's a typical scenario:
- host creates a time to meet (or event on a website) for another user or group of users
- host gives other users a password after they've registered for the event
- host can start a session with an optional password
- other users can see the session publicly on the session plate
- when any user tries to enter the session, they are asked for the password and can only join with the correct password

---
Additionally, it would be great to have this connected to a world (file) and not just sessions. This would allow us to share a world, but only individuals with the password would be able to open it and host a session (which maybe also has the password protection by default). But perhaps this could be a separate related request.

**Comment by Frooxius** [comment]
This could be added, although I'm not sure if it's the best approach to solve this problem. Password can easily be leaked at which point you have no control over who joins. 

What about sending invites to users without having to add them to friends list instead and making one-time use invites that could be used externally? This might provide better user experience as well as better security and control.

**Comment by sirkitree** [comment]
I think one time links would solve this quite nicely, yes.

**Comment by ProbablePrime** [comment]
I agree that passwords are not secure and have extensive experience of this being the case. 

Expanding our invitation capabilities in a similar manner to Discord's system(limited use etc) and combining that with enhancements to permissions seems like a better move.

**Comment by sirkitree** [comment]
I've fleshed out this idea a bit more here: https://paper.dropbox.com/doc/Neos-Events--BCS4wmgLRUaeCXTJe46DMyp9AQ-MwnBBGKcRoypkPB9AXK96

**Comment by Frooxius** [comment]
Thanks for putting that together! There's quite a lot there, that goes beyond just this single feature. There's a few major ones that I can see:

- The store / marketplace system for publishing paid worlds / access and items
- Event system integration (e.g. EventBrite) that requires some form of account tie in
- Neos' native event organizing system, which is something that's planned to be added. I'd like to build as much on top of this as possible, but integrations with 3rd party will also be important for business applications
- Custom Neos builds? This will require the custom build system and launcher/updater to be done first. However we'd prefer them to use the regular Neos build in most cases and just use a thin launcher, which will pass the event information and startup info (e.g. to load simplified dash that you customize, forced avatar and so on).
- The world configuration would probably happen in the world itself though as part of building it, rather than through the publishing process. You'd need to test it as well, so it's better if it's consistent that way.

A lot of the stuff we already have some underlying technical stuff for that we can build off for this (e.g. the customizable dash / UI, we did use the system for the Metamovie project, with custom facets designed for the event), some of it will have to be built first. Doing the custom launchers/builds, with customized UI and such would definitely be a Business/Pro feature. There's other things to consider too (hosting of the sessions) that might be provided as service to make it easier for organizers.

But a lot of it goes beyond the scope of this issue and needs multiple things to be implemented first.

**Comment by sirkitree** [comment]
Thanks :) I posted the above not as a means of creating additional scope to this issue, but to establish that password protected sessions would be very useful for these use-cases.

I wonder if there has been any additional progress on this issue though, as the use cases I posted in that document are something we've found more folks looking to implement. Is there a better place for tracking that additional scope (some of which I have heard is somewhat in the works)?

**Comment by shiftyscales** [comment]
There are mechanisms that have been recently introduced that can cover some of the needs of this issue (cloud variables) which also were added to assist with the purpose of performances/events like The Metamovie Project, and VBLFC. @sirkitree

https://wiki.neos.com/Cloud_Variables#Access

Effectively, anyone that you wish to have access to the 'password' you would set-up in advance by assigning them a allowUserCloudVariable.

This would also solve the proposed security issue Frooxius mentioned of the password possibly leaking.
